### DISCUSSION TITLE
Please add support for the srt protocol (Secure Reliable Transport)

### DISCUSSION THREAD

**Discussion by gargolito** [PyAV-Org/PyAV#1503] [2. Feature Ideas]
## Overview

FFMpeg has support to fully utilize the [SRT protocol](https://github.com/Haivision/srt) 

## Existing FFmpeg API

[libavformat/libsrt.c](https://ffmpeg.org/doxygen/trunk/libsrt_8c.html)

## Expected PyAV API

Example:
```
container = av.open(srt://ipaddress:port?mode=caller)
```
This will currently fail with "unable to open file"

## Investigation
Tried to open an srt video stream in caller mode
```
container = av.open(srt://IPADDRESS:PORT/?mode=caller)
```

## Reproduction
1. Start a listener (this is the _server_): `ffmpeg -hide_banner -loglevel error -re -stream_loop -1 -i SOME_VIDEO_FILE -c copy -map 0  -f mpegts srt://localhost:5765/?mode=listener`
2.  try to open the stream with python script using pyav as usual

## Versions

- OS: ubuntu linux 22.04
- PyAV runtime: 12.3.0
```
PyAV v12.3.0
library configuration: --disable-static --enable-shared --libdir=/tmp/vendor/lib --prefix=/tmp/vendor --disable-alsa --disable-doc --disable-libtheora --disable-mediafoundation --enable-fontconfig --enable-gmp --enable-gnutls --enable-libaom --enable-libbluray --enable-libdav1d --enable-libfreetype --enable-libmp3lame --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenjpeg --enable-libopus --enable-libspeex --enable-libtwolame --enable-libvorbis --enable-libvpx --enable-libwebp --enable-libxcb --enable-libxml2 --enable-lzma --enable-zlib --enable-version3 --enable-libopenh264 --disable-libx264
library license: LGPL version 3 or later
libavcodec     60. 31.102
libavdevice    60.  3.100
libavfilter     9. 12.100
libavformat    60. 16.100
libavutil      58. 29.100
libswresample   4. 12.100
libswscale      7.  5.100
```
- PyAV build:
pip installed, got version 
``` 
v12.3.0
````
- FFmpeg:
```
ffmpeg version 4.4.2-0ubuntu0.22.04.1 Copyright (c) 2000-2021 the FFmpeg developers
built with gcc 11 (Ubuntu 11.2.0-19ubuntu1)
configuration: --prefix=/usr --extra-version=0ubuntu0.22.04.1 --toolchain=hardened --libdir=/usr/lib/x86_64-linux-gnu --incdir=/usr/include/x86_64-linux-gnu --arch=amd64 --enable-gpl --disable-stripping --enable-gnutls --enable-ladspa --enable-libaom --enable-libass --enable-libbluray --enable-libbs2b --enable-libcaca --enable-libcdio --enable-libcodec2 --enable-libdav1d --enable-libflite --enable-libfontconfig --enable-libfreetype --enable-libfribidi --enable-libgme --enable-libgsm --enable-libjack --enable-libmp3lame --enable-libmysofa --enable-libopenjpeg --enable-libopenmpt --enable-libopus --enable-libpulse --enable-librabbitmq --enable-librubberband --enable-libshine --enable-libsnappy --enable-libsoxr --enable-libspeex --enable-libsrt --enable-libssh --enable-libtheora --enable-libtwolame --enable-libvidstab --enable-libvorbis --enable-libvpx --enable-libwebp --enable-libx265 --enable-libxml2 --enable-libxvid --enable-libzimg --enable-libzmq --enable-libzvbi --enable-lv2 --enable-omx --enable-openal --enable-opencl --enable-opengl --enable-sdl2 --enable-pocketsphinx --enable-librsvg --enable-libmfx --enable-libdc1394 --enable-libdrm --enable-libiec61883 --enable-chromaprint --enable-frei0r --enable-libx264 --enable-shared
libavutil      56. 70.100 / 56. 70.100
libavcodec     58.134.100 / 58.134.100
libavformat    58. 76.100 / 58. 76.100
libavdevice    58. 13.100 / 58. 13.100
libavfilter     7.110.100 /  7.110.100
libswscale      5.  9.100 /  5.  9.100
libswresample   3.  9.100 /  3.  9.100
libpostproc    55.  9.100 / 55.  9.100
```

## Additional context

I really just need to be able to open srt streams, so that basic functionality would be enough for me.

**Comment by JoeSchiff** [comment]
@gargolito 
I got you

Let me know if you have any advice for writing tests for this.

**Reply by gargolito** [reply]
I looked at the issue and you already did what I would have done to get a minimum viable product. There may be other tests, but I don't see it is necessary now that you've incorporated libsrt.

**Reply by JoeSchiff** [reply]
Sorry about the delay. There is some problem with our windows runner which has made building anything impossible. I can't implement srt until we get it resolved

**Comment by gargolito** [comment]
Understood. Thanks. (I don't use windows because of... reasons😉)

Best Regards,

Anthony Camilo

**Comment by JoeSchiff** [comment]
@gargolito

Would it be a problem if we were to bundle srt without support for encryption? Building without openssl would simplify things.

(I also don't use windows)

**Comment by gargolito** [comment]
I need encryption for my use case. I'm perfectly comfortable compiling it if you can give me adhoc instructions for adding support for encryption.

**Reply by JoeSchiff** [reply]
It looks like your ffmpeg has libsrt enabled. I believe if you were to install pyav with `pip install av --no-binary av` it should use the ffmpeg that you already have installed. 

I'll be able to test this more tomorrow

**Reply by JoeSchiff** [reply]
This seems to work on my system. I have ffmpeg with --enable-libsrt installed through apt.

**Reply by gargolito** [reply]
Are you saying that pyav works with srt as it is?

**Reply by JoeSchiff** [reply]
Yes, I believe so. You just need a version of ffmpeg with --enable-libsrt, which it seems you have. 

However, if you run `pip install av` then it will install a version of ffmpeg without --enable-libsrt. 

Try out `pip install av --no-binary av` and let me know if it works for you.

**Reply by JoeSchiff** [reply]
PyAV 14.1.0 has been released and includes libsrt.

**Reply by gargolito** [reply]
Tested on my end with haivision srtgateway using encrypted stream and it works great. Thanks!

**Comment by gargolito** [comment]
Awesome, thanks for the update!

Best Regards,

Anthony Camilo
